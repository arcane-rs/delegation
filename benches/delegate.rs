use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
