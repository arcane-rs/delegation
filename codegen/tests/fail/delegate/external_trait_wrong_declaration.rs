use delegation::delegate;

#[delegate(as = AsRef)]
trait AsRefDef<T: ?Sized> {
    fn as_ref(&mut self) -> &mut T;
}

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

#[delegate(as = AsStr)]
trait AsStrDef {
    fn as_str(&self) -> &str;
}

#[delegate(derive(
    AsRef<str> as AsRefDef,
    AsStr as AsStrDef,
))]
enum Name {
    First(String),
}

fn main() {}
