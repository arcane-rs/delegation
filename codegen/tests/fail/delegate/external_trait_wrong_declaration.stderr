error[E0053]: method `as_ref` has an incompatible type for trait
 --> tests/fail/delegate/external_trait_wrong_declaration.rs:5:15
  |
5 |     fn as_ref(&mut self) -> &mut T;
  |               ^^^^^^^^^ types differ in mutability
  |
  = note: expected signature `fn(&__delegate_AsRefDef__Wrapper<__Delegate>) -> &T`
             found signature `fn(&mut __delegate_AsRefDef__Wrapper<__Delegate>) -> &mut T`
help: change the self-receiver type to match the trait
  |
5 |     fn as_ref(&self) -> &mut T;
  |               ~~~~~

error[E0053]: method `as_ref` has an incompatible type for trait
  --> tests/fail/delegate/external_trait_wrong_declaration.rs:5:15
   |
5  |       fn as_ref(&mut self) -> &mut T;
   |                 ^^^^^^^^^ types differ in mutability
...
24 | / #[delegate(derive(
25 | |     AsRef<str> as AsRefDef,
26 | |     AsStr as AsStrDef,
27 | | ))]
   | |___- in this procedural macro expansion
   |
   = note: expected signature `fn(&Name) -> &str`
              found signature `fn(&mut Name) -> &mut str`
   = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: change the self-receiver type to match the trait
   |
5  |     fn as_ref(&self) -> &mut T;
   |               ~~~~~

error[E0308]: mismatched types
 --> tests/fail/delegate/external_trait_wrong_declaration.rs:3:1
  |
3 | #[delegate(as = AsRef)]
  | ^^^^^^^^^^^^^^^^^^^^^^^ types differ in mutability
4 | trait AsRefDef<T: ?Sized> {
5 |     fn as_ref(&mut self) -> &mut T;
  |                             ------ expected `&'__delegate mut T` because of return type
  |
  = note: expected mutable reference `&'__delegate mut T`
                     found reference `&T`
  = note: this error originates in the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/fail/delegate/external_trait_wrong_declaration.rs:3:1
   |
3  |   #[delegate(as = AsRef)]
   |   ^^^^^^^^^^^^^^^^^^^^^^^
   |   |
   |   types differ in mutability
   |   expected `&mut str` because of return type
...
24 | / #[delegate(derive(
25 | |     AsRef<str> as AsRefDef,
26 | |     AsStr as AsStrDef,
27 | | ))]
   | |___- in this procedural macro expansion
   |
   = note: expected mutable reference `&mut str`
                      found reference `&str`
   = note: this error originates in the macro `AsRefDef` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
